File: ios-native/NeobankNative/Sources/Features/Auth/AuthRepository.swift
import Foundation

/// Wraps the two auth-adjacent clients: `authClient` for the unauthenticated
/// `/v1/auth/login` and `/v1/auth/register` endpoints, `client` for calls that
/// need (and may transparently refresh) a bearer token.
struct AuthRepository: Sendable {
    let authClient: APIClient
    let client: APIClient

    func login(email: String, password: String) async throws -> AuthTokens {
        try await authClient.send(
            "/v1/auth/login",
            method: .post,
            body: ["email": email, "password": password]
        )
    }

    func register(email: String, password: String, phone: String?, inviteCode: String?) async throws -> AuthTokens {
        try await authClient.send(
            "/v1/auth/register",
            method: .post,
            body: [
                "email": email,
                "password": password,
                "phone": phone?.isEmpty == false ? phone : nil,
                "invite_code": inviteCode?.isEmpty == false ? inviteCode : nil,
            ]
        )
    }

    func profile() async throws -> Profile {
        try await client.send("/v1/me")
    }

    func changePassword(current: String, new: String) async throws {
        let _: EmptyResponse = try await client.send(
            "/v1/auth/change-password",
            method: .post,
            body: ["current_password": current, "new_password": new]
        )
    }
}

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/auth/AuthRepository.kt
package dev.horobets.neobanknative.features.auth

import dev.horobets.neobanknative.core.networking.ApiClient
import dev.horobets.neobanknative.core.networking.EmptyResponse
import dev.horobets.neobanknative.core.networking.HttpMethod
import dev.horobets.neobanknative.core.networking.jsonBodyOf

/**
 * Wraps the two auth-adjacent clients: [authClient] for the unauthenticated
 * `/v1/auth/login` and `/v1/auth/register` endpoints, [client] for calls
 * that need (and may transparently refresh) a bearer token.
 */
class AuthRepository(
    private val authClient: ApiClient,
    private val client: ApiClient,
) {
    suspend fun login(email: String, password: String): AuthTokens = authClient.send(
        "/v1/auth/login",
        method = HttpMethod.POST,
        body = jsonBodyOf("email" to email, "password" to password),
    )

    suspend fun register(email: String, password: String, phone: String?, inviteCode: String?): AuthTokens = authClient.send(
        "/v1/auth/register",
        method = HttpMethod.POST,
        body = jsonBodyOf(
            "email" to email,
            "password" to password,
            "phone" to phone?.ifEmpty { null },
            "invite_code" to inviteCode?.ifEmpty { null },
        ),
    )

    suspend fun profile(): Profile = client.send("/v1/me")

    suspend fun changePassword(current: String, new: String) {
        client.send<EmptyResponse>(
            "/v1/auth/change-password",
            method = HttpMethod.POST,
            body = jsonBodyOf("current_password" to current, "new_password" to new),
        )
    }
}
